### PR TITLE
Adds a variety of contact lenses 

### PR DESCRIPTION
**Pull request by Cisrox**
Basically, you can wear prescription contacts that act as prescription glasses, and polarized contact lenses that serve as sunglasses, without the glasses sprites covering your eyes.

You can also change your eye color.

:cl:
 * rscadd: Adds prescription contact lenses, and polarized contact lenses that serve as sunglasses.
 * rscadd: Adds nanite contact lenses that you can use to dye your eyes.
 * rscadd: Adds a contact lens crate to cargo.

### CODE REVIEW

File: code/game/objects/items/weapons/cosmetics.dm [hunk 1]
```diff
@@ -491,3 +491,71 @@
...
+/obj/item/weapon/nanitecontacts/New()
+	..()
+	color_r = rand(0,255)
+	color_g = rand(0,255)
+	color_b = rand(0,255)
+	update_icon()
+
+/obj/item/weapon/nanitecontacts/update_icon()
+	overlays.len = 0
+	var/icon/contact_color = new/icon("icon" = 'icons/obj/items.dmi', "icon_state" = "contacts_overlay")
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm]: Make this create a new `image()` with the icon, set its color variable and add that as overlay instead.

`/icon` operations are ludicrously expensive and they're not necessary here.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 2]
```diff
@@ -491,3 +491,71 @@
...
+	contact_color.Blend(rgb(color_r, color_g, color_b), ICON_ADD)
+	overlays += contact_color
+
+/obj/item/weapon/nanitecontacts/attack_self(mob/user as mob)
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm]: Remove `as mob`

File: code/game/objects/items/weapons/cosmetics.dm [hunk 3]
```diff
@@ -491,3 +491,71 @@
...
+	var/new_color = input(user, "Choose the contact's color:", "Color Select") as color|null
+	if(new_color)
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm]: This should check whether the user is still alive and still has the contacts in their hands.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 4]
```diff
@@ -491,3 +491,71 @@
...
+		color_r = hex2num(copytext(new_color, 2, 4))
+		color_g = hex2num(copytext(new_color, 4, 6))
+		color_b = hex2num(copytext(new_color, 6, 8))
+	update_icon()
+
+/obj/item/weapon/nanitecontacts/attack(mob/M as mob, mob/user as mob)
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm]: Remove `as mob`

File: code/game/objects/items/weapons/cosmetics.dm [hunk 5]
```diff
@@ -491,3 +491,71 @@
...
+	if(!istype(M, /mob))
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm]: You can leave out the second argument since BYOND infers it.

**Review by PJB3005 — DISMISSED** [review]

File: code/datums/supplypacks.dm [hunk 6]
```diff
@@ -647,6 +647,20 @@ var/list/all_supply_groups = list("Supplies","Clothing","Security","Hospitality"
 	containername = "russian clothing crate"
 	group = "Clothing"
 	contraband = 1
+
+/datum/supply_packs/contacts
+	name = "Contact Lenses"
+	contains = list(/obj/item/clothing/glasses/contacts/polarized,
+					/obj/item/clothing/glasses/contacts/polarized,
```
Comment (Kurfursten) [inline code/datums/supplypacks.dm]: I know cargo *can* get shades by ordering secvend, building a vendor, hacking it, and buying the shades, but you rarely see that because it's a pain compared to just ordering a crate, unlike glubbs.

Not to mention these are much stealthier than shades, right?

Comment (coderbus13) [reply code/datums/supplypacks.dm]: dusty crate has shades in it

**Comment by Dacendeth** [comment]
I assume these will look bad on vox like almost every other eyewear?

File: code/datums/supplypacks.dm [hunk 7]
```diff
@@ -647,6 +647,20 @@ var/list/all_supply_groups = list("Supplies","Clothing","Security","Hospitality"
 	containername = "russian clothing crate"
 	group = "Clothing"
 	contraband = 1
+
+/datum/supply_packs/contacts
+	name = "Contact Lenses"
```
Comment (DamianX) [inline code/datums/supplypacks.dm]: Lenses vs lenses

File: code/game/objects/items/weapons/cosmetics.dm [hunk 8]
```diff
@@ -491,3 +491,71 @@
 		return
 	if (prob(25))
 		shatter()
+
+/obj/item/weapon/nanitecontacts
+	name = "Nanite Contacts"
```
Comment (DamianX) [inline code/game/objects/items/weapons/cosmetics.dm]: lowercase

File: code/game/objects/items/weapons/cosmetics.dm [hunk 9]
```diff
@@ -491,3 +491,71 @@
...
+	update_icon()
+
+/obj/item/weapon/nanitecontacts/attack(mob/M as mob, mob/user as mob)
+	if(!istype(M, /mob))
+		return
+
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		var/area = user.zone_sel.selecting
+		var/area_string = "eyes"
```
Comment (DamianX) [inline code/game/objects/items/weapons/cosmetics.dm]: Why is this necessary? Is this ever going to be applied to anything but eyes? I guess just replace occurrences of this with "eyes"

File: code/modules/clothing/glasses/glasses.dm [hunk 10]
```diff
@@ -314,3 +314,18 @@
 	icon_state = "kaminaglasses"
 	item_state = "kaminaglasses"
 	cover_hair = 1
+
+/obj/item/clothing/glasses/contacts
+	name = "Contact Lenses"
```
Comment (DamianX) [inline code/modules/clothing/glasses/glasses.dm]: Contact Lenses vs contact lenses

File: code/modules/clothing/glasses/glasses.dm [hunk 11]
```diff
@@ -314,3 +314,18 @@
...
+	desc = "Only nerds wear glasses."
+	icon = 'icons/obj/items.dmi'
+	icon_state = "contact"
+	prescription = 1
+
+/obj/item/clothing/glasses/contacts/polarized
+	name = "Polarized Contact Lenses"
```
Comment (DamianX) [inline code/modules/clothing/glasses/glasses.dm]: make this lowercase

File: code/modules/clothing/glasses/glasses.dm [hunk 12]
```diff
@@ -314,3 +314,18 @@
 	icon_state = "kaminaglasses"
 	item_state = "kaminaglasses"
 	cover_hair = 1
+
+/obj/item/clothing/glasses/contacts
+	name = "Contact Lenses"
+	desc = "Only nerds wear glasses."
+	icon = 'icons/obj/items.dmi'
+	icon_state = "contact"
+	prescription = 1
```
Comment (DamianX) [inline code/modules/clothing/glasses/glasses.dm]: use TRUE

File: code/modules/clothing/glasses/glasses.dm [hunk 13]
```diff
@@ -314,3 +314,18 @@
...
+
+/obj/item/clothing/glasses/contacts/polarized
+	name = "Polarized Contact Lenses"
+	desc = "Protects your eyes from bright flashes of light."
+	icon_state = "polarized_contact"
+	darkness_view = -1
+	prescription = 1
```
Comment (DamianX) [inline code/modules/clothing/glasses/glasses.dm]: use TRUE

**Review by DamianX — DISMISSED** [review]

**Comment by Riekland** [comment]
can these be made to not work on greys because they're roughly a quarter of the size of their eyes

File: code/modules/clothing/glasses/glasses.dm (hunk 12 above)
Comment (Cisrox) [reply code/modules/clothing/glasses/glasses.dm]: all other forms of glasses use 1 instead of true

**Comment by Cisrox** [comment]
@Dacendeth they don't have a sprite when they're equipped, so your eyes just show naturally

@Riekland that makes sense but greys suffer the worst from wearing glasses, they don't even fit over their eyes

File: code/game/objects/items/weapons/cosmetics.dm (hunk 1 above)
Comment (Cisrox) [reply code/game/objects/items/weapons/cosmetics.dm]: how to do this

Comment (Cisrox) [reply code/game/objects/items/weapons/cosmetics.dm]: also help with checking if user is alive and still has contacts in their hand

**Comment by Riekland** [comment]
Having invisible flash prot is a bad idea
Maybe make it so that you can emag them to engage the shielding or have them be an R&D item?

**Comment by Cisrox** [comment]
you can already make sunglasses invisible using a gas mask

**Comment by Lokiamis** [comment]
Do they block eye jabs?

**Comment by Cisrox** [comment]
i'll make them not block eye jabs, and also move the polarized ones into rnd

File: code/game/objects/items/weapons/cosmetics.dm (hunk 1 above)
Comment (PJB3005) [reply code/game/objects/items/weapons/cosmetics.dm]: Basically you want to do

```dm
var/image/I = image(icon = 'icons/obj/items.dmi', icon_state = "contacts_overlay")
I.color = rgb(color_r, color_g, color_b)
overlays += I
```
Might not produce the same blending as with an icon but eh.

**Review by jknpj — DISMISSED** [review]
should be fine

File: code/modules/clothing/glasses/glasses.dm [hunk 14]
```diff
@@ -314,3 +314,19 @@
...
 	item_state = "kaminaglasses"
 	cover_hair = 1
+
+/obj/item/clothing/glasses/contacts
+	name = "contact lenses"
+	desc = "Only nerds wear glasses."
+	icon = 'icons/obj/items.dmi'
+	icon_state = "contact"
+	prescription = 1
+	body_parts_covered = ~EYES
```
Comment (ghost) [inline code/modules/clothing/glasses/glasses.dm]: Doesn't this cover literally every body part except the eyes?

Comment (jknpj) [reply code/modules/clothing/glasses/glasses.dm]: didn't saw this but yes you need to use &= ~EYES

Comment (Kurfursten) [reply code/modules/clothing/glasses/glasses.dm]: you could do that in the /New() but I am pretty sure &= won't work in a definition because it's an operation, definitions need constant expressions

Comment (jknpj) [reply code/modules/clothing/glasses/glasses.dm]: T-thanks BYOND

Comment (Cisrox) [reply code/modules/clothing/glasses/glasses.dm]: i tried compiling with &= and it failed

Comment (ghost) [reply code/modules/clothing/glasses/glasses.dm]: literally just = EYES if you want to cover the eyes

Comment (Kurfursten) [reply code/modules/clothing/glasses/glasses.dm]: He wants the opposite, he wants it to not cover the eyes (because they're contacts)

In this case though you just want `body_parts_covered = null` to override the inherited eye covering

Comment (Cisrox) [reply code/modules/clothing/glasses/glasses.dm]: thanks i'll commit this in a second

**Comment by Cisrox** [comment]
my byond compiler stopped working so I think I need to reinstall the repository to test this, once I do that it should be ready

**Comment by ShiftyRail** [comment]
@Cisrox Give a bit more details, I'm 99% sure you don't have to reclone the repo

**Comment by Cisrox** [comment]
when i try to compile about three thousands errors pop up and it fails compiling 
but the travis check works so theres nothing wrong with the code
